### ISSUE TITLE
Assertion failed '(effOp1->gtOper == GT_CNS_INT) && (effOp1->IsIntegralConst(0) || effOp1->IsIntegralConst(1))'

### ISSUE THREAD

**Issue by v-haren** [dotnet/runtime#33091]
failed in job: [runtime-coreclr jitstress 20200302.1](https://dev.azure.com/dnceng/public/_build/results?buildId=543560&view=ms.vss-test-web.build-test-results-tab&runId=17211062&resultId=107192&paneView=debug)

related test:
JIT/Methodical/Boxing/misc/_relnestval_il/_relnestval_il.sh
JIT/Methodical/Boxing/callconv/_relinstance_il/_relinstance_il.sh

Error message
`
\nAssert failure(PID 71955 [0x00011913], Thread: 2245291 [0x2242ab]): Assertion failed '(effOp1->gtOper == GT_CNS_INT) && (effOp1->IsIntegralConst(0) || effOp1->IsIntegralConst(1))' in 'BoxTest.Test:FibonacciImpl(System.Object):System.Object:this' during 'Assertion prop' (IL size 383)\n\n File: /Users/runner/runners/2.165.0/work/1/s/src/coreclr/src/jit/morph.cpp Line: 10798\n Image: /private/tmp/helix/working/B1B509B1/p/corerun\n\n/private/tmp/helix/working/B1B509B1/w/B6AC0A1B/e/JIT/Methodical/Boxing/misc/_relnestval_il/_relnestval_il.sh: line 275: 71955 Abort trap: 6 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"\n\nReturn code: 1\nRaw output file: /private/tmp/helix/working/B1B509B1/w/B6AC0A1B/e/JIT/Methodical/Reports/JIT.Methodical/Boxing/misc/_relnestval_il/_relnestval_il.output.txt\nRaw output:\nBEGIN EXECUTION\n/tmp/helix/working/B1B509B1/p/corerun _relnestval_il.dll ''\nExpected: 100\nActual: 134\nEND EXECUTION - FAILED\nTest Harness Exitcode is : 1\nTo run the test:\n> set CORE_ROOT=/tmp/helix/working/B1B509B1/p\n> /private/tmp/helix/working/B1B509B1/w/B6AC0A1B/e/JIT/Methodical/Boxing/misc/_relnestval_il/_relnestval_il.sh\nExpected: True\nActual: False
`

**Comment by BruceForstall** [comment]
@briansull Can you investigate?
cc @dotnet/jit-contrib

**Comment by benaadams** [comment]
```
C:\h\w\AE3808BB\w\A461086C\e>C:\h\w\AE3808BB\p\CoreRun.exe C:\h\w\AE3808BB\p\xunit.console.dll JIT\Methodical\JIT.Methodical.XUnitWrapper.dll -parallel collections -nocolor -noshadow -xml testResults.xml -trait TestGroup=JIT.Methodical.a-dA-D 
Microsoft.DotNet.XUnitConsoleRunner v2.5.0 (64-bit .NET Core 5.0.0-ci)
  Discovering: JIT.Methodical.XUnitWrapper
  Discovered:  JIT.Methodical.XUnitWrapper
  Starting:    JIT.Methodical.XUnitWrapper
    JIT\Methodical\Boxing\callconv\_relinstance_il\_relinstance_il.cmd [FAIL]
      
      Assert failure(PID 5092 [0x000013e4], Thread: 6036 [0x1794]): 
Assertion failed '(effOp1->gtOper == GT_CNS_INT) && (effOp1->IsIntegralConst(0) || effOp1->IsIntegralConst(1))' in 'BoxTest.Test:Fibonacci2(System.Object,System.Object):System.Object:this' during 'Assertion prop' (IL size 137)
      
          File: F:\workspace\_work\1\s\src\coreclr\src\jit\morph.cpp Line: 10798
          Image: C:\h\w\AE3808BB\p\CoreRun.exe
      
      
      Return code:      1
      Raw output file:      C:\h\w\AE3808BB\w\A461086C\e\JIT\Methodical\Reports\JIT.Methodical\Boxing\callconv\_relinstance_il\_relinstance_il.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\h\w\AE3808BB\p\corerun.exe" _relinstance_il.dll 
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=C:\h\w\AE3808BB\p
      > C:\h\w\AE3808BB\w\A461086C\e\JIT\Methodical\Boxing\callconv\_relinstance_il\_relinstance_il.cmd
      Expected: True
      Actual:   False
      Stack Trace:
        F:\workspace\_work\1\s\artifacts\tests\coreclr\Windows_NT.x64.Checked\TestWrappers\JIT.Methodical\JIT.Methodical.XUnitWrapper.cs(15369,0): at JIT_Methodical._Boxing_callconv__relinstance_il__relinstance_il_._Boxing_callconv__relinstance_il__relinstance_il_cmd()
      Output:
        
        Assert failure(PID 5092 [0x000013e4], Thread: 6036 [0x1794]): 
Assertion failed '(effOp1->gtOper == GT_CNS_INT) && (effOp1->IsIntegralConst(0) || effOp1->IsIntegralConst(1))' in 'BoxTest.Test:Fibonacci2(System.Object,System.Object):System.Object:this' during 'Assertion prop' (IL size 137)
        
            File: F:\workspace\_work\1\s\src\coreclr\src\jit\morph.cpp Line: 10798
            Image: C:\h\w\AE3808BB\p\CoreRun.exe
        
        
        Return code:      1
        Raw output file:      C:\h\w\AE3808BB\w\A461086C\e\JIT\Methodical\Reports\JIT.Methodical\Boxing\callconv\_relinstance_il\_relinstance_il.output.txt
        Raw output:
        BEGIN EXECUTION
         "C:\h\w\AE3808BB\p\corerun.exe" _relinstance_il.dll 
        Expected: 100
        Actual: -1073740286
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=C:\h\w\AE3808BB\p
        > C:\h\w\AE3808BB\w\A461086C\e\JIT\Methodical\Boxing\callconv\_relinstance_il\_relinstance_il.cmd
    JIT\Methodical\Boxing\misc\_relnestval_il\_relnestval_il.cmd [FAIL]
```

**Comment by briansull** [comment]
investigating

**Comment by briansull** [comment]
Repros with:
COMPLUS_JitDump=Fibonacci2
COMPLUS_JitStress=2

**Comment by briansull** [comment]
@erozenfeld 

Dump ends with:

```
VN based non-null prop in BB08:
N002 (  4,  3) [000319] ---X--------              *  NULLCHECK byte   $108

```

**Comment by briansull** [comment]
```
04 0000002e`147f96d0 00007ff9`689b21d8 clrjit!noWayAssertBody+0x3c [d:\fxkit\runtime\src\coreclr\src\jit\error.cpp @ 527] 
05 0000002e`147f9700 00007ff9`68b7cdcf clrjit!noWayAssertBodyConditional+0x48 [d:\fxkit\runtime\src\coreclr\src\jit\error.cpp @ 514] 
06 0000002e`147f9730 00007ff9`68b87d3a clrjit!Compiler::fgMorphSmpOp+0x44f [d:\fxkit\runtime\src\coreclr\src\jit\morph.cpp @ 10797] 
07 0000002e`147f9fd0 00007ff9`68b6b59c clrjit!Compiler::fgMorphTree+0x52a [d:\fxkit\runtime\src\coreclr\src\jit\morph.cpp @ 14364] 
08 0000002e`147fa200 00007ff9`689400ea clrjit!Compiler::fgMorphBlockStmt+0xdc [d:\fxkit\runtime\src\coreclr\src\jit\morph.cpp @ 15080] 
09 0000002e`147fa2d0 00007ff9`68936254 clrjit!Compiler::optVNAssertionPropCurStmt+0x12a [d:\fxkit\runtime\src\coreclr\src\jit\assertionprop.cpp @ 5120] 
0a 0000002e`147fa360 00007ff9`68979220 clrjit!Compiler::optAssertionPropMain+0x194 [d:\fxkit\runtime\src\coreclr\src\jit\assertionprop.cpp @ 5168] 
```

**Comment by briansull** [comment]
```
fgMorphTree BB05, STMT00034 (before)
               [000202] --CXG-------              *  JTRUE     void  
               [000201] --CXG-------              \--*  EQ        int   
               [000197] --CX--------                 +--*  INTRINSIC ref    objGetType
               [000196] ------------                 |  \--*  LCL_VAR   ref    V16 tmp10        
               [000200] --C-G-------                 \--*  CALL help ref    HELPER.CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
               [000198] ------------ arg0               \--*  CNS_INT(h) long   0xd1ffab1e class
Optimizing compare of obj.GetType() and type-from-handle to compare method table pointer
GenTreeNode creates assertion:
               [000315] ---X--------              *  NULLCHECK byte  
fgMorphTree BB05, STMT00034 (after)
               [000317] ---XG-------              *  JTRUE     void  
               [000318] ---X--------              \--*  COMMA     int   
               [000319] ---X--------                 +--*  NULLCHECK byte  
               [000320] ------------                 |  \--*  LCL_VAR   ref    V16 tmp10        
               [000321] ------------                 \--*  CNS_INT   int    1
```

**Comment by briansull** [comment]
Problem is cause by the insertion of a GT_NOT node:

```
Decided to reverse conditional branch at block BB10 branch to BB12 since it falls into a rarely run block
Relocated rarely run block BB11 by reversing conditional jump at BB10

N006 (  8,  7) [000317] ---XG-------              *  JTRUE     void  
N005 (  6,  5) [000530] ---X--------              \--*  NOT       int   
N004 (  5,  4) [000318] ---X--------                 \--*  COMMA     int   
N002 (  4,  3) [000319] ---X--------                    +--*  NULLCHECK byte  
N001 (  3,  2) [000320] ------------                    |  \--*  LCL_VAR   ref    V16 tmp10        
N003 (  1,  1) [000321] ------------                    \--*  CNS_INT   int    1
```

**Comment by briansull** [comment]
Fixed by  #33337
